docs/developers/adr/004_magnetic_lattice.md

```diff
@@ -0,0 +1,22 @@
+# Classes for different kinds of magnetic structures
+
+## Context
+
+SpinW has different methods for dealing with different kinds of simulation, some simulations only work for systems decribed by a single (possibly complex) propagation vector, some (e.g. "biquadratic") only work
+when the propagation vector is (0,0,0). It seems that a useful distinction is between systems where every unique magnetic state is explicitly represented, and ones where it is implicitly defined. 
+The notion of commensurability maps on to this. Commensurate propagation vectors permit a finite (though potentially large) explicit representation.
+
+The plan is to make separate classes for these descriptions, provide some degree of conversion between the two, and associate appropriate calculations with each one.
+1) The explicit structure - every unique site is described explicity - explcit supercell, only (0,0,0), supports biquadratic
+2) The implicit structure - decribed by a unit cell, and a propagation vector - implicit supercell, supports incommensuate and true helical structures. The implcit structure requires the "trick" used in spinw to calculate intensities at +/- the propagation vector.
```

Comment:
I like the separation of the two types of structure given the calculation proceeds differently for the two. I'm assuming (1) would also support k=0 with explicit supercell of `nExt = [1,1,1]` - I don't know if you would call this explicit supercell as it is the default.

I'm less sure about the names (but I can't come up with anything better!). By the SpinW definition we could call them (1) Commensurate and (2) Incommensurate. However as discussed this can cause confusion: a modulation of (1/2, 0, 0) with a single nuclear/parent unit cell would be classed as incommensurate in SpinW despite being a commensurate modulation by the common definition!

Reply:
In terms of names, what about `SuperCell` and `RotatingFrame`?

We could still have a `create_magnetic_structure` function for users similar to the current `genmagstr` method which would create a `SuperCell` or `RotatingFrame` depending on its input, or we can separate out into different helper functions, like `create_helical_structure`, `create_fourier_structure`, `create_random_structure`, etc for the different [modes](https://github.com/SpinW/spinw/blob/73f604ab4d84084d872d1f5fdf46dbc54e14cdd7/swfiles/%40spinw/genmagstr.m#L51-L115) of `genmagstr`.

Reply:
In terms of conversion between the two classes, you should always be able to convert from the implicit to the explicit representation (just construct a supercell with size determined by the inverse of the propagation vector).

However, conversion from an explicit to implicit representation will not always be possible. We could write a check for this (it would involve checking that the phase angle between spins in adjacent base unit cell is the same) but it might be easier to disallow this conversion as I think that probably only a minority of explicit cases can be converted to an implicit representation.